File: backend/job-scheduler-rust/src/repositories/bill_repository.rs
use sqlx::{MySql, Pool};

use crate::models::bill_model::{Bill, CreateBill};

pub struct BillRepository<'a> {
    pub pool: &'a Pool<MySql>,
}

impl<'a> BillRepository<'a> {
    pub fn new(pool: &'a Pool<MySql>) -> Self {
        Self { pool }
    }

    pub async fn get_by_name_and_due_date(
        &self,
        name: &String,
        user: &String,
        due_date: &chrono::NaiveDate,
    ) -> Result<Option<Bill>, sqlx::Error> {
        let bills = sqlx::query_as::<_, Bill>(
            r#"
            SELECT
                id,
                name,
                amount,
                due_date,
                status,
                is_generated_by_recurrence,
                user,
                modified_at,
                created_at
            FROM
                bills
            WHERE
                name = ? AND user = ? AND due_date = ?
            "#,
        )
        .bind(name)
        .bind(user)
        .bind(due_date)
        .fetch_optional(self.pool)
        .await?;

        Ok(bills)
    }

    pub async fn create_bill(&self, data: CreateBill) -> Result<u64, sqlx::Error> {
        let result = sqlx::query(
            r#"
            INSERT INTO bills (name, amount, due_date, status, user, is_generated_by_recurrence)
            VALUES (?, ?, ?, ?, ?, ?)
            "#,
        )
        .bind(data.name)
        .bind(data.amount)
        .bind(data.due_date)
        .bind(data.status)
        .bind(data.user)
        .bind(data.is_generated_by_recurrence)
        .execute(self.pool)
        .await?;

        Ok(result.last_insert_id())
    }
}

File: backend/core-api/src/repositories/billRepository.ts
import { ResultSetHeader, RowDataPacket } from "mysql2";
import pool from "../config/db";
import { IBill, IBillDBSchema } from "../shared/interfaces/IBill";

export class BillRepository {
  userId: number;

  constructor(userId: number) {
    this.userId = userId;

    this.get = this.get.bind(this);
    this.checkBillExist = this.checkBillExist.bind(this);
    this.create = this.create.bind(this);
    this.updateMetadata = this.updateMetadata.bind(this);
    this.delete = this.delete.bind(this);
  }

  async get(onlyOverdue: boolean): Promise<IBillDBSchema[]> {
      let SQLQuery = "SELECT * FROM bills WHERE user_id = ?"

      if(onlyOverdue) SQLQuery += " AND due_date < CURRENT_DATE";

      const [rows] = await pool.query(SQLQuery, [this.userId]);

      return rows as IBillDBSchema[];
  }

  async checkBillExist(billId: number): Promise<boolean> {
    const SQLQuery = "SELECT * FROM bills WHERE user_id = ? AND id = ?";

    const [rows] = await pool.query<RowDataPacket[]>(SQLQuery, [this.userId, billId]);

    return rows.length > 0;
  }

  async create({ name, amount, dueDate, status, isRecurring }: IBill): Promise<boolean> {
    const SQLQuery = `
    INSERT INTO bills (name, amount, due_date, status, is_generated_by_recurrence, user_id)
    VALUES (?, ?, ?, ?, ?, ?)
  `;

    try {
      const [result] = await pool.query<ResultSetHeader>(SQLQuery, [
        name,
        amount,
        dueDate,
        status ?? "",
        isRecurring ? 1 : 0,
        this.userId
      ]);

      return result.affectedRows > 0;
    } catch (error) {
      return false;
    }
  }

  async updateMetadata(
    billId: number,
    fieldsToUpdate: string[],
    metadataValues: (string | number)[]
  ): Promise<boolean> {
    try {
      const SQLUpdateQuery = `UPDATE bills SET ${fieldsToUpdate.join(", ")} WHERE id = ? AND user_id = ?`;

      const [result] = await pool.query<ResultSetHeader>(SQLUpdateQuery, [...metadataValues, billId, this.userId]);

      return result.affectedRows > 0;
    } catch (error) {
      return false;
    }
  }

  async delete(billId: number): Promise<boolean> {
    try {
      const SQLQuery = `DELETE FROM bills WHERE id = ? AND user_id = ?`;

      const [result] = await pool.query<ResultSetHeader>(SQLQuery, [billId, this.userId]);

      return result.affectedRows > 0;
    } catch (error) {
      return false;
    }
  }
}
